### DISCUSSION TITLE
LaTeX-like figure placement in Typst

### DISCUSSION THREAD

**Discussion by jack-davison** [quarto-dev/quarto-cli#8801] [Q&A]
In Typst via Quarto, is it possible to do the equivalent of LaTeX figures being [b]/[t]?

LaTeX is very good at being able to just put a figure somewhere sensible near enough to where you've put it in the source code, but without leaving a load of whitespace in the document.

I've found Typst in quarto is, by default, a lot more like Word, where if your plot is a bit too big it'll be nudged onto the next page, leaving a big gap at the bottom of the previous one. Ideally I'd want the text after the figure to flow into that gap, which to me is nicer typesetting. 

Can this be done in Typst?

**Comment by mcanouil** [comment]
Regarding the option for positioning, Typst support is at its beginning (within Pandoc and Quarto).

Regarding the "float" part of your question, it's unrelated to Quarto.
I don't have the answer, but it definitely is on <https://typst.app/docs/>, which is the best place to learn about what Typst features are.

**Comment by cderv** [comment]
I wonder if this is still a limitation of Typst 
https://typst.app/docs/guides/guide-for-latex-users/#limitations

> **Page break optimization.** LaTeX runs some smart algorithms to not only optimize line but also page breaks. While Typst tries to avoid widows and orphans, it uses less sophisticated algorithms to determine page breaks. You can insert custom page breaks in Typst using #pagebreak(weak: true) before submitting your document. The argument weak ensures that no double page break will be created if this spot would be a natural page break anyways. You can also use #v(1fr) to distribute space on your page. It works quite similar to LaTeX's \vfill.

**Comment by jack-davison** [comment]
I see - more of a Typst issue than a Quarto one!

As a total Typst newbie, I thought having access to the "placement" arg through Quarto would be the fix, but perhaps this wouldn't do what I'm after after all!

```
#figure(
  placement: top,  // or `auto` or `bottom`
  caption: [A Floating Figure],
  image("figure.svg", width: 80%),
)
```

**Reply by cderv** [reply]
🤔 this is interesting indeed. You add `keep-typ` in your project to have the `typ` file and then play in https://typst.app/ to have live preview of your document. I think the placement of figures produced by quarto is associated to how the alignment is done. 

I don't know typst enough to know if this would help or not, but if you find that working and useful customization we may consider improving this for future version. This could also a be a Pandoc feature request as some of the figures Typst code is generated by Pandoc directly - so not just a quarto thing IMO. 

Please to test ! And feedback from user is always appreciated ! Thanks

**Comment by tpoisot** [comment]
I have experimented a bit with 

~~~
include-before-body:
      - text: |
        #set figure(placement: auto) 
~~~

but this seems to have no effect - I think that for manuscript preparation (and conference papers), having figures that float in a sensical way is important, so I'm keen to help experiment on this some more!

**Comment by mattharrison** [comment]
I'm not using quarto, but have been experimenting with this syntax:

```
![Second image](img/alta-2.png){width=50% placement=bottom}
```

My pandoc filter will add the placement attribute to the Figure, but it appears that Pandoc ignores that in the final typst output.

**Reply by cderv** [reply]
> but it appears that Pandoc ignores that in the final typst output.

If you can work on a reproducible example with Pandoc directly, it should probably be reported in Pandoc repo. 
If this is only happening with Quarto, you can open a new issue sharing what your filter is doing and what quarto does not do as you would expect.
